design/feature-flags.md

```diff
@@ -0,0 +1,92 @@
+# Feature Flags
+
+Sometimes we want to introduce new breaking behavior because we believe this is
+the correct default behavior for the CDK. The problem, of course, is that
+breaking changes are only allowed in major versions and those are rare.
+
+This document describes a proposal for a pattern/mechanism called feature flags.
+It will allow us to introduce breaking behavior which is disabled by default (so
+existing projects will not be affected) but enabled automatically for new
+projects created through `cdk init`.
+
+## Approach
+
+The basic idea is that new breaking behavior will always be disabled by default
+and only enabled when a certain CDK context parameter is set. If not enabled,
+the system will continue to behave exactly like it used to without breaking any
+existing projects.
+
+When we release a new major version of the AWS CDK, we will flip this behavior
+or completely remove the legacy behavior.
+
+In order for new projects to pick up this new behavior automatically, we will
+modify `cdk init` to inject the set of "future flags" into the generated
+`cdk.json` file. This means that the new project will have the latest behavior,
+but projects that were created prior to the introduction of this feature will
+have the same legacy behavior based on the set of capabilities that were
+available at the time of the project's creation. This list will be cleaned up
+every time we release a major version of course.
```

Comment:
what'll be the strategy for communicating the flags that can be established for any given version? Will we just vend it through documentation? CLI command to get the list of flags supported? other?

Reply:
See "Future Developments". If this proliferates, we may end up with something like a `cdk features` command, but I wouldn't go and implement this before it's actually an issue.